### PR TITLE
fix: isolate asynchronous inference lifecycles

### PR DESCRIPTION
**Pull request by ryjen**
## Summary

Implements the first safe slice of #62 by making JavaScript inference reusable and single-flight across synchronous and asynchronous entry points.

- creates a fresh observable and accumulated-output buffer for every hook generation
- tracks one active generation per hook and settles it exactly once
- observes controller-level cancellation and close lifecycle events
- uses layout-effect teardown so unmount cancellation cannot emit completion
- returns and uses request-owned cancellation handles in the example
- enforces one active operation across every `LlmPipe` sharing the same native module
- rejects sync/async overlap with typed `GenerationInProgressError`
- prevents non-owner controllers from closing shared active work
- defers owner close until synchronous generation finishes
- scopes native event listeners to an internal generation ID so stale callbacks are ignored
- removes listeners after final, error, cancellation, close, or native startup failure
- claims context-inference ownership before asynchronous retrieval begins
- prevents inference from starting after unmount or retrieval-order races

## Root cause

`useInferenceAsync()` previously retained one observable for the hook lifetime. The observable completed after the first final event but retained its accumulated text closure, making later calls unreliable. At the controller layer, globally named native events and a shared subscription collection allowed overlapping requests, cancellation, and close operations to interfere with each other.

## Concurrency contract

This slice deliberately uses **single-flight per native module identity**. An overlapping synchronous or asynchronous request is rejected with `GenerationInProgressError` rather than cancelling or queueing the active request. The guard also applies when separate `LlmPipe` instances share React Native's singleton native module.

Parallel native generation remains out of scope until backend memory, session, and callback behavior are explicitly supported.

## Tests

Regression coverage includes:

- sequential generations and output reset
- callback-driven generation after terminal events
- synchronous/asynchronous overlap in both directions
- overlap across pipes sharing one native module
- external controller cancellation followed by another request
- request-owned cancellation in the example
- cancellation ownership and stale-event suppression
- non-owner close protection
- deferred owner close during synchronous generation
- terminal and startup-failure cleanup
- unmount cleanup without completion notification
- unmount during pending context retrieval
- concurrent context-retrieval ordering
- callback-free terminal cleanup
- structured error inheritance and codes

## Follow-up scope

This PR does not complete the native correlation portion of #62. Android and iOS still emit globally named string events. Follow-up work must add request IDs to the native method/event contract, normalize partial/final semantics, and verify native lifecycle and cancellation parity.

## Validation

All required checks pass on head `2856edc56901be90fe1a6ff027a697fbda823ed2`:

- CI: lint, TypeScript, unit tests, components package, Android example, and iOS example
- Coverage Gate
- Node Compatibility Matrix
- CodeQL

All inline review threads are resolved.

Refs #62

### CODE REVIEW

File: src/Amaryllis.ts [hunk 1]
```diff
@@ -48,12 +54,22 @@ export class LlmPipe implements LlmEngine {
     params: LlmRequestParams,
     callbacks?: LlmCallbacks
   ): Promise<void> {
-    if (callbacks) {
-      this.setupAsyncCallbacks(callbacks);
+    if (activeNativeGenerations.has(this.llmNative)) {
+      throw new GenerationInProgressError();
     }
```
Comment (ryjen) [inline src/Amaryllis.ts]: **P1 — enforce single-flight across both generation APIs.** This guard exists only in `generateAsync()`, while `generate()` above can still enter the same shared native engine/session during an active stream. The issue's contract is single-flight per native engine/session, and the Android implementation uses the same mutable `LlmInferenceSession` for both methods. Reject synchronous generation while an async request owns the module (and vice versa), or explicitly prove/support safe multiplexing.

File: src/Amaryllis.ts [hunk 2]
```diff
@@ -62,23 +78,28 @@ export class LlmPipe implements LlmEngine {
   }
```
Comment (ryjen) [inline src/Amaryllis.ts:126]: **P1 — do not let a non-owner close shared active work.** With two `LlmPipe` instances sharing the singleton native module, this pipe can have `activeGenerationId === null` while another pipe owns the WeakMap entry. `cancelAsync()` then does nothing, but `llmNative.close()` destroys the other pipe's engine; that owner's listeners and global lock are never released. Track the owning pipe/request in the shared registry and make `close()` settle/release the actual owner before closing.

**Review by ryjen — COMMENTED** [review]
Blocking findings:

1. The existing P1 thread is valid: cancelling through the public controller removes `LlmPipe` listeners but never settles `useInferenceAsync`'s `activeGenerationRef`, so the mounted hook permanently rejects later prompts. The example uses this exact path.
2. The shared-native-module lock protects only `generateAsync`; `generate()` can enter the same native engine/session while an async generation is active, violating the documented single-flight contract.
3. A non-owning `LlmPipe.close()` still closes the shared native module. If another pipe owns the active generation, its listeners and WeakMap lock remain live while the underlying engine is destroyed, leaving future generations stuck.
4. `useContextInferenceAsync` performs asynchronous augmentation before the base hook owns a generation. If the component unmounts while `engine.search()` is pending, the continuation can start native inference after hook cleanup has already run. Two overlapping context requests can also race so the later invocation acquires the slot first.

The existing green matrix does not cover these paths: the ChatPrompt integration test mocks `useContextInferenceAsync`, and lifecycle tests cancel only through the returned hook function after generation has started.

File: src/Amaryllis.ts [hunk 3]
```diff
@@ -40,45 +58,88 @@ export class LlmPipe implements LlmEngine {
...
+    const operationId = this.claimNativeOperation('sync');
+    try {
+      const nativeParams = toNativeRequestParams(params);
+      return await this.llmNative.generate(nativeParams);
+    } finally {
+      this.releaseNativeOperation(operationId);
+      if (this.closeRequested) {
+        this.close();
+      }
+    }
```
Comment (ryjen) [inline src/Amaryllis.ts:81]: **P2 — preserve the generation error when a deferred close also fails.** This `close()` call runs inside `generate()`'s `finally`. If native generation rejects and the deferred native `close()` throws, the close exception replaces the original inference failure, hiding the actionable root cause from callers. Wrap the deferred close separately (matching provider cleanup behavior), or retain/rethrow the generation error while reporting the close failure. Add coverage for generation rejection plus a throwing `close()`.

**Review by ryjen — COMMENTED** [review]
One remaining correctness edge case should be fixed before merge. The request-scoped lifecycle, cancellation observation, shared-operation ownership, context ordering, tests, and current CI otherwise look sound.
